lib/vector-core/src/event/metric/tags.rs

```diff
@@ -1,14 +1,283 @@
+    /// Convert this set into a single value, mimicking the behavior of this set being just a plain
+    /// single string while still storing all of the values.
+    pub(crate) fn into_single(self) -> Option<String> {
+        match self {
+            Self::Single(tag) => tag,
+            Self::Set(set) => set.into_iter().last(),
+        }
+    }
+
+    /// Get the "single" value of this set, mimicking the behavior of this set being just a plain
+    /// single string while still storing all of the values.
+    pub(crate) fn as_single(&self) -> Option<&str> {
+        match self {
+            Self::Single(tag) => tag.as_ref().map(String::as_str),
+            Self::Set(set) => set.iter().last().map(String::as_ref),
+        }
+    }
+
+    fn iter(&self) -> <&Self as IntoIterator>::IntoIter {
+        self.into_iter()
+    }
+
+    fn is_empty(&self) -> bool {
+        match self {
+            Self::Single(None) => true,
+            Self::Single(Some(_)) | Self::Set(_) => false, // the `Set` variant will never be empty
+        }
+    }
+
+    #[cfg(test)]
+    fn len(&self) -> usize {
+        match self {
+            Self::Single(None) => 0,
+            Self::Single(Some(_)) => 1,
+            Self::Set(set) => set.len(),
+        }
+    }
+
+    #[cfg(test)]
+    fn contains(&self, value: &str) -> bool {
+        match self {
+            Self::Single(None) => false,
+            Self::Single(Some(tag)) => tag == value,
+            Self::Set(set) => set.contains(value),
+        }
+    }
+
+    fn retain(&mut self, mut f: impl FnMut(&str) -> bool) {
+        match self {
+            Self::Single(None) => (),
+            Self::Single(Some(tag)) => {
+                if !f(tag) {
+                    *self = Self::Single(None);
+                }
+            }
+            Self::Set(set) => {
+                set.retain(|value| f(value.as_str()));
+                match set.len() {
+                    0 | 1 => *self = Self::Single(set.pop()),
+                    _ => {}
+                }
+            }
+        }
+    }
+
+    fn insert(&mut self, value: TagValue) -> bool {
+        match self {
+            // Need to take ownership of the single value to optionally move it into a set.
+            Self::Single(single) => match single.take() {
+                None => {
+                    *self = Self::Single(Some(value));
+                    false
+                }
+                Some(tag) => {
+                    if tag == value {
+                        *self = Self::Single(Some(value));
+                        true
+                    } else {
+                        *self = Self::Set(IndexSet::from([tag, value]));
+                        false
+                    }
+                }
+            },
+            Self::Set(set) => {
+                // If the value was previously present, we want to move it to become the last element. The
+                // only way to do this is to remove any existing value.
```

Comment:
To avoid changing behavior, we need the last assigned value to become the "single" value. We extract that "single" value from the last element in the set. Ergo, when re-adding the same element, it has to become the last element. I could add some logic to skip re-adding if it is already the last element, though.

Reply:
Gotcha, makes sense now. 👍🏻